src/token_adapter/jwt.ts

```
@@ -1,57 +1,125 @@
         if (error) {
           reject(error);
         } else {
           resolve(<string> jwt);
         }
+
         return;
       });
     });
   }
 
-  private _decode(token: string): BluebirdPromise<any> {
-    return new BluebirdPromise((resolve, reject) => {
-      if (token === undefined) {
-        resolve();
-      } else {
-        verify(token, secret, (error, decodedToken) => {
-          if (error) {
-            reject(new Error('invalid token'));
-          } else {
-            resolve(<ITokenData> decodedToken);
-          }
-        });
-      }
+  private _decode(token: string, secretOrPublicKey: string): Promise<ITokenData> {
+    return new Promise((resolve: any, reject: any): any => {
+      verify(token, secretOrPublicKey, (error: any, decodedToken: any): void => {
+        if (error) {
+          reject(new Error('invalid token'));
+        } else {
+          resolve(<ITokenData> decodedToken);
+        }
+
+        return;
+      });
     });
   }
 
+  private async _getSecretOrPrivateKey(algorithm: string): Promise<string> {
+    switch (algorithm) {
+      case 'RS256':
+        const key: string = this.config && this.config.key ? this.config.key : undefined;
+        if (key === undefined) {
+          return undefined;
+        }
+
+        return this.pkiService.getPrivateKey(key);
+
+      case 'HS256':
+        // default, if no config exists is fallback for backward compatibility, it should be removed
+        const secret: string = this.config && this.config.secret ? this.config.secret : 'secret';
+
+        return secret;
+
+      default:
+        return undefined;
+    }
+  }
+
+  private async _getSecretOrPublicKey(algorithm: string): Promise<string> {
+    switch (algorithm) {
+      case 'RS256':
+        const key: string = this.config && this.config.key ? this.config.key : undefined;
+        if (key === undefined) {
+          return undefined;
```

`undefined` is not a public key. This should throw an error.